### PR TITLE
Tee mmu refactoring

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jbech-linaro** [comment]
Just skimmed through it, looks mostly like moving code and renaming. I like the shortened names.
`Acked-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by jenswi-linaro** [comment]
> Just skimmed through it, looks mostly like moving code and renaming.

Thanks, that's a good summary. :-)

**Comment by jforissier** [comment]
LGTM too. Perhaps the first commit subject "core: tee_mmu.h: remove tee_mmu_map_add_segment()" could be "core: tee_mmu.h: remove **unused** tee_mmu_map_add_segment()".

`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Comment addressed and tags applied. Hopefully all build errors fixed too.

File: core/arch/arm/mm/core_mmu.c [hunk 1]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+
+	/* extra check: we could rely on  core_mmu_get_mem_by_type() */
+	if (!tee_vbuf_is_non_sec(s, e - s))
+		panic("PUB RAM is not non-secure");
+
+#ifdef CFG_PL310
+	/* Allocate statically the l2cc mutex */
+	tee_l2cc_store_mutex_boot_pa(virt_to_phys((void *)s));
+	s += sizeof(uint32_t);			/* size of a pl310 mutex */
+	s =  ROUNDUP(s, SMALL_PAGE_SIZE);	/* keep required alignment */
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2310]: minor: 2 space chars

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
 	return p->pa;
 }
+
+#ifdef CFG_CORE_RESERVED_SHM
+static TEE_Result teecore_init_pub_ram(void)
+{
+	vaddr_t s = 0;
+	vaddr_t e = 0;
+
+	/* get virtual addr/size of NSec shared mem allcated from teecore */
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2296]: s/allcated/allocated/

File: core/arch/arm/mm/core_mmu.c [hunk 3]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+	core_mmu_get_mem_by_type(MEM_AREA_NSEC_SHM, &s, &e);
+
+	if (s >= e || s & SMALL_PAGE_MASK || e & SMALL_PAGE_MASK)
+		panic("invalid PUB RAM");
+
+	/* extra check: we could rely on  core_mmu_get_mem_by_type() */
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2302]: 2 space chars

File: core/arch/arm/mm/core_mmu.c [hunk 4]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+#endif /*CFG_CORE_RESERVED_SHM*/
+
+void core_mmu_init_ta_ram(void)
+{
+	vaddr_t s = 0;
+	vaddr_t e = 0;
+	paddr_t ps = 0;
+	paddr_t pe = 0;
+
+	/* get virtual addr/size of RAM where TA are loaded/executedNSec
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2328]: fix coding style
```
/*
 * ...
 */

File: core/arch/arm/mm/core_mmu.c [hunk 5]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+	 * shared mem allcated from teecore */
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2329]: s/allcated/allocated/

File: core/arch/arm/mm/core_mmu.c [hunk 6]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+	virt_get_ta_ram(&s, &e);
+#endif
+	ps = virt_to_phys((void *)s);
+	pe = virt_to_phys((void *)(e - 1)) + 1;
+
+	if (!ps || (ps & CORE_MMU_USER_CODE_MASK) ||
+	    !pe || (pe & CORE_MMU_USER_CODE_MASK))
+		panic("invalid TA RAM");
+
+	/* extra check: we could rely on  core_mmu_get_mem_by_type() */
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2342]: 2 space chars

File: core/arch/arm/mm/core_mmu.c [hunk 7]
```diff
@@ -2287,3 +2286,68 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
...
+	if (!tee_pbuf_is_sec(ps, pe - ps))
+		panic("TA RAM is not secure");
+
+	if (!tee_mm_is_empty(&tee_mm_sec_ddr))
+		panic("TA RAM pool is not empty");
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:2347]: should be fixed:
```patch
-		panic("TA RAM pool is not empty");
+		panic("TA RAM pool is empty");
```

File: core/tee/tee_svc_cryp.c [hunk 8]
```diff
@@ -3250,11 +3249,10 @@ TEE_Result syscall_authenc_update_aad(unsigned long state,
...
 
-	res = tee_mmu_check_access_rights(&to_user_ta_ctx(sess->ctx)->uctx,
-					  TEE_MEMORY_ACCESS_READ |
-					  TEE_MEMORY_ACCESS_ANY_OWNER,
-					  (uaddr_t) aad_data,
-					  aad_data_len);
+	res = vm_check_access_rights(&to_user_ta_ctx(sess->ctx)->uctx,
+				     TEE_MEMORY_ACCESS_READ |
+				     TEE_MEMORY_ACCESS_ANY_OWNER,
+				     (uaddr_t) aad_data, aad_data_len);
```
Comment (etienne-lms) [inline core/tee/tee_svc_cryp.c:3255]: remove space char at type cast.

File: core/tee/tee_svc_storage.c [hunk 9]
```diff
@@ -285,10 +285,8 @@ TEE_Result syscall_storage_obj_open(unsigned long storage_id, void *object_id,
 		goto exit;
 	}
 
-	res = tee_mmu_check_access_rights(&utc->uctx,
-					  TEE_MEMORY_ACCESS_READ,
-					  (uaddr_t) object_id,
-					  object_id_len);
+	res = vm_check_access_rights(&utc->uctx, TEE_MEMORY_ACCESS_READ,
+				     (uaddr_t) object_id, object_id_len);
```
Comment (etienne-lms) [inline core/tee/tee_svc_storage.c:289]: remove space char at type cast.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with some minor issues (typo, space chars) from original code that could (should) be fixed.

File: core/arch/arm/mm/core_mmu.c (hunk 7 above)
Comment (jenswi-linaro) [reply core/arch/arm/mm/core_mmu.c:2347]: That would be incorrect.

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c:2347]: Sorry, you're right. Actually the error message looks strange. It's the allocation reference list  in the pool that is not empty (and should be), not the pool itself. But that's not related to this change.
Discard my comment.

**Comment by jenswi-linaro** [comment]
Comments addressed and tag applied.

**Pull request merged**